Build of HPX fails when find_package(Boost) is called before FetchContent_MakeAvailable(hpx)

## Expected Behavior

Build succeeds

## Actual Behavior

Build fails because of misconfiguration of HPX's use of filesystem.

```
[  7%] Building CXX object _deps/hpx-build/libs/logging/CMakeFiles/hpx_logging.dir/src/logging.cpp.o                                             
  In file included from /home/kor/development/objects/Debug/lue/_deps/hpx-src/libs/logging/src/logging.cpp:11:0:                                   
  /home/kor/development/objects/Debug/lue/_deps/hpx-src/libs/filesystem/include/hpx/modules/filesystem.hpp:22:10: fatal error: filesystem: No such 
   #include <filesystem>                                                                                                                           
            ^~~~~~~~~~~~                                                                                                                           
  compilation terminated.                                                                                                                          
  _deps/hpx-build/libs/logging/CMakeFiles/hpx_logging.dir/build.make:75: recipe for target '_deps/hpx-build/libs/logging/CMakeFiles/hpx_logging.dir
  make[2]: *** [_deps/hpx-build/libs/logging/CMakeFiles/hpx_logging.dir/src/logging.cpp.o] Error 1                                                 
  make[2]: *** Waiting for unfinished jobs....
```

## Steps to Reproduce the Problem

  1. Setup a project that builds HPX using CMake's FetchContent module
  1. Call `find_package(Boost REQUIRED COMPONENTS filesystem)` before calling the FetchContent stuff
  1. Build project

One way to do this is to clone https://github.com/kordejong/hello_hpx and add `find_package(Boost REQUIRED COMPONENTS filesystem)` to this line: https://github.com/kordejong/hello_hpx/blob/da5a4347bd4f036793b8ecf39bff58a4814b14d3/CMakeLists.txt#L87 and `bash ./debug_build.sh`.

## Specifications

  - HPX Version: Today's master, but I have seen this issue some weeks/months ago too
  - Platform (compiler, OS): Kubuntu 18.04, gcc-7

## Remarks
- In `HPX_SetupBoostFilesystem.cmake`, some configuration logic is conditionally executed, based on whether `Boost::filesystem` is defined. If the parent project has already found Boost filesystem, the target `Boost::filesystem` is defined and HPX is not configured correctly.
- As a user of HPX I can work around this issue by switching the find_package(Boost) and FetchContent stuff around. First build HPX, then find Boost filesystem ourselves.
- The issue is subtle and not easy to figure out. Maybe it would be good if `HPX_SetupBoostFilesystem.cmake` would work correctly even if find_package(Boost) has been called before the build of HPX. Maybe it is just a matter of moving the `hpx_add_config_define_namespace` outside of the conditional?

--- a/cmake/HPX_SetupBoostFilesystem.cmake
+++ b/cmake/HPX_SetupBoostFilesystem.cmake
@@ -7,13 +7,19 @@
 if(HPX_FILESYSTEM_WITH_BOOST_FILESYSTEM_COMPATIBILITY)
   # In case find_package(HPX) is called multiple times
   if(NOT TARGET Boost::filesystem)
-    hpx_add_config_define_namespace(
-      DEFINE HPX_FILESYSTEM_HAVE_BOOST_FILESYSTEM_COMPATIBILITY
-      NAMESPACE FILESYSTEM
-    )
-
     find_package(Boost ${Boost_MINIMUM_VERSION} MODULE COMPONENTS filesystem)
+
+    if(NOT Boost_FILESYSTEM_FOUND)
+      hpx_error(
+        "Could not find Boost.Filesystem Provide a boost installation including the filesystem library"
+      )
+    endif()
   endif()
+
+  hpx_add_config_define_namespace(
+    DEFINE HPX_FILESYSTEM_HAVE_BOOST_FILESYSTEM_COMPATIBILITY
+    NAMESPACE FILESYSTEM
+  )
 else()
   if(NOT HPX_WITH_CXX17_FILESYSTEM)
     hpx_error(

--- a/cmake/HPX_SetupBoostProgramOptions.cmake
+++ b/cmake/HPX_SetupBoostProgramOptions.cmake
@@ -5,23 +5,22 @@
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
 if(HPX_PROGRAM_OPTIONS_WITH_BOOST_PROGRAM_OPTIONS_COMPATIBILITY)
-  set(__boost_program_options Boost::program_options)
-endif()
+  if(NOT TARGET Boost::program_options)
+    find_package(
+      Boost ${Boost_MINIMUM_VERSION} MODULE COMPONENTS program_options
+    )
 
-if(HPX_PROGRAM_OPTIONS_WITH_BOOST_PROGRAM_OPTIONS_COMPATIBILITY
-   AND NOT TARGET Boost::program_options
-)
+    if(NOT Boost_PROGRAM_OPTIONS_FOUND)
+      hpx_error(
+        "Could not find Boost.ProgramOptions. Provide a boost installation including the program_options library"
+      )
+    endif()
+  endif()
+
+  set(__boost_program_options Boost::program_options)
 
   hpx_add_config_define_namespace(
     DEFINE HPX_PROGRAM_OPTIONS_HAVE_BOOST_PROGRAM_OPTIONS_COMPATIBILITY
     NAMESPACE PROGRAM_OPTIONS
   )
-
-  find_package(Boost ${Boost_MINIMUM_VERSION} MODULE COMPONENTS program_options)
-
-  if(NOT Boost_PROGRAM_OPTIONS_FOUND)
-    hpx_error(
-      "Could not find Boost.ProgramOptions. Provide a boost installation including the program_options library"
-    )
-  endif()
 endif()

--- a/tests/unit/build/fetchcontent/CMakeLists.txt
+++ b/tests/unit/build/fetchcontent/CMakeLists.txt
@@ -11,6 +11,10 @@
 cmake_minimum_required(VERSION 3.13)
 project(HPXFetchContent LANGUAGES CXX)
 
+# Test that HPX correctly handles having Boost Filesystem and ProgramOptions
+# found before setting up HPX, if they are available (do not REQUIRE these)
+find_package(Boost COMPONENTS filesystem program_options)
+
 include(FetchContent)
 
 fetchcontent_declare(

